readme.md
# @m4rch/sv

this is a package for file-serving with automatically added live-reload functionality

# usage

## cli

```
$ sv --help
  Description
    quickly serve and live-reload a folder for development.

  Usage
    $ sv [dir] [options]

  Options
    -p, --port        specify the port  (default 4000)
    -o, --open        open the website in a browser
    -s, --static      disable live-reloading and content updates
    -d, --list-dir    always list directories
    -v, --version     Displays current version
    -h, --help        Displays this message

  Examples
    $ sv dist
    $ sv -p 5000
```

## api

you can import the serve function via

```js
import serve from "@m4rch/sv"
```

the serve function takes an object as the only argument.

```ts
interface options {
	dir: string // the directory you want to serve
	port: number // the port of the wished service
	wsPort: number // the port of the live-reload socket
	static?: boolean // whether or not to live-reload,
	listDir?: boolean // whether or not to always list directories
}
```

# credit

the ui is heavily inspired by the [sirv-cli](https://github.com/lukeed/sirv)


changelog.md
# changelog

all notable changes to this project will be documented in this file.

the format is loosely based on [keep a changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [semantic versioning](https://semver.org/spec/v2.0.0.html).

## [unreleased]

## [0.1.6]

### added

- typescript typings via jsdoc syntax

## [0.1.5]

### added

- added a second port option 3000 to defaults

### changed

- changed changelog.md

## [0.1.4] - 2021-08-15

### added

- added --list-dir / -d option
- added repository to package.json
- bumped ws version

### fixed

- fixed time zone offset in logs

## [0.1.3] - 2021-07-10

### added

- added comment to injected live-reload code

### changed

- updated version for cli

### fixed

- fixed breaking bug in static mode

## [0.1.2] - 2021-07-09

### added

- added static mode (--static flag)
- sorted files while serving directory

### changed

- improved directory design while serving directory
- improved readme.md
- added default values for sade

### fixed

- fixed bug when serving directory with index.html
- fixed bug not reloading when file removed
- fixed import error (package.json)
- fixed potential bug with the coloring

## [0.1.1] - 2021-07-01

### added

- directories now show the files contained within it

### changed

- updated readme.md
- removed uuid module
- moved the handler

### fixed

- fixed bug that caused the main entry directory to not update on file change
- removed spelling mistake from readme.md

## [0.1.0] - 2021-06-28

### added

- initial release


src/extra.js
export const scripts = {
	/**
	 * generate code to inject for live-reload
	 * @param {number} wsPort port for live reload websocket
	 * @returns {string}
	 */
	livereload: ( wsPort ) => [
		"<script>",
		"\t/* code injected by sv for live-reload purposes */",
		"\t(function () {",
		`\t\tconst ws = new WebSocket(\"ws://localhost:${wsPort}\")`,
		"\t\tws.addEventListener(\"message\", ({ data }) => data == \"reload\" && location.reload())",
		"\t}())",
		"</script>",
		""
	].join("\n")
}

export const styles = {
	directory: [
		"<style>",
		"* {",
		"\tmargin: 0;",
		"}",
		".name {",
		"\tpadding-bottom: 15px;",
		"\tmargin-bottom: 20px;",
		"\tborder-bottom: 2px solid #cccccc;",
		"\tfont-family: monospace;",
		"\ttext-align: center;",
		"}",
		".name a {",
		"\tcolor: #3579bd;",
		"\tfont-size: 25px;",
		"\tfont-weight: 900;",
		"\ttext-decoration: none;",
		"}",
		".name a:hover {",
		"\tcolor: #3579bd;",
		"\ttext-decoration: underline;",
		"}",
		".main {",
		"\tmargin: 20px 25%;",
		"\tpadding: 20px;",
		"\tborder: 1px solid #cccccc;",
		"}",
		".entries > a {",
		"\tdisplay: block;",
		"\tpadding: 3px 10px;",
		"\tcolor: #3579bd;",
		"\tfont-size: 17px;",
		"\tfont-family: monospace;",
		"\ttext-decoration: none;",
		"}",
		".entries > a:hover {",
		"\ttext-decoration: underline;",
		"}",
		"</style>"
	].join("")
}


src/index.js
import { createServer } from "http"
import { parse, send, log, is404 } from "./utils.js"
import { init as livereload } from "./livereload.js"
import { watch, recurse } from "./routes.js"
import html from "./html.js"

/**
 * serve file
 * @param {import("./private.js").ServeOptions} options options
 */
export default function serve ({ dir, port, wsPort, static: statik = false, listDir = false }) {
	const server = createServer(async ( req, res ) => {
		const { path } = parse(req)
		const render = await html({ dir, path, wsPort, statik, listDir })

		if (render.code != 404) {
			send({ res, content: render.content, headers: render.headers })
		} else {
			is404(res)
		}

		log({ code: render.code, path })
	})

	if (!statik) {
		livereload(wsPort)
		watch(dir)
	} else {
		recurse(dir, [])
	}

	server.listen(port)
}


src/utils.js
import * as colors from "kleur/colors"
import { exec } from "child_process"
import { randomBytes } from "crypto"

/**
 * parse url
 * @param {import("node:http").IncomingMessage} message 
 * @returns 
 */
export function parse ({ url }) {
	const [ path = "/" ] = decodeURI(/** @type {string} */ (url)).split("?")
	return {
		raw: url,
		path: path.length > 1 && path.endsWith("/") ? path.slice(0, -1) : path
	}
}

/**
 * send 404 request
 * @param {import("node:http").ServerResponse} res 
 */
export function is404 ( res ) {
	res.writeHead(404)
	res.end()
}

/**
 * send response
 * @param {{ res: import("node:http").ServerResponse, content: Buffer, headers: Record<string, string | number> }} param0 
 */
export function send ({ res, content, headers }) {
	res.writeHead(200, headers)
	res.end(content)
}

/**
 * generate headers
 * @param {{ type: string, content: Buffer }} options options
 * @returns {Record<string, string | number>}
 */
export function toHeaders ({ type, content }) {
	return {
		"Content-Length": content.length,
		"Content-Type": type || "",
		"Cache-Control": "no-cache"
	}
}

/**
 * logging
 * @param {{ code: number, path: string }} options options
 */
export function log ({ code, path }) {
	const date = /** @type {string} */ (new Date(Date.now() - new Date().getTimezoneOffset() * 60_000).toISOString().split(/[T\.]/g)[1])
	const dateString = colors.magenta(date)
	const fn = code >= 400 ? "red" : code > 300 ? "yellow" : "green"

	console.log(`  [${dateString}] ${colors[fn](code)} - ${path}`)
}

/**
 * open in browser
 * @param {number} port port
 */
export function launch ( port ) {
	const url = `http://localhost:${port}`

	if (process.platform == "win32") {
		exec(`start ${url}`)
	} else if (process.platform == "linux") {
		exec(`xdg-open ${url}`)
	} else if (process.platform == "darwin") {
		exec(`open ${url}`)
	}
}

export function random () {
	return randomBytes(16).toString("hex")
}


src/routes.js
import CheapWatch from "cheap-watch"
import { lstat, readFile, readdir } from "fs/promises"
import { existsSync } from "fs"
import { join } from "path"
import { update } from "./livereload.js"

/** @type {Map<string, import("./private.js").Route>} */
export const routes = new Map

/**
 * watch files
 * @param {string} dir dir
 */
export function watch ( dir ) {
	const watcher = new CheapWatch({ dir, debounce: 100 })
	watcher.init()

	watcher.on("+", async ({ path, stats }) => {
		const file = join(dir, path)

		const routeString = `/${file}`
		const route = await getChildren(path, stats)
		routes.set(routeString, route)

		update()
	})

	watcher.on("-", ({ path }) => {
		if (routes.has(`/${path}`)) {
			routes.delete(`/${path}`)
			update()
		}
	})
}

/**
 * 
 * @param {string} dir directory
 * @param {string} file file path
 * @returns {Promise<boolean>}
 */
export async function get ( dir, file ) {
	const path = join(dir, file)
	if (!existsSync(path)) return false

	const stats = await lstat(path)

	const routeString = `/${file}`
	const route = await getChildren(path, stats)
	routes.set(routeString, route)

	return true
}

/**
 * recurse directory
 * @param {string} dir directory
 * @param {string[]} route route
 */
export async function recurse ( dir, route = [] ) {
	get(dir, route.join("/"))

	const children = await readdir(join(dir, route.join("/")))
	children.forEach(async ( file ) => {
		const stats = await lstat(join(dir, route.join("/"), file))

		if (stats.isDirectory()) {
			recurse(dir, route.concat(file))
		} else {
			get(dir, route.concat(file).join("/"))
		}
	})
}

/**
 * get content
 * @param {string} path path
 * @param {import("node:fs").Stats} stats stats
 * @returns {Promise<import("./private.js").Route>}
 */
async function getChildren ( path, stats ) {
	if (stats.isDirectory()) {
		let children = await readdir(path)
		let nchildren = (await Promise.all(
			children
				.sort(( el1, el2 ) => el1.toLowerCase().localeCompare(el2.toLowerCase()))
				.map(async ( p ) => ({ path: p, directory: (await lstat(join(path, p))).isDirectory() }))
		)).sort(( el1, el2 ) => +el2.directory - +el1.directory)

		return { directory: true, children: nchildren }
	} else {
		const content = await readFile(path)
		return { directory: false, content }
	}
}


src/html.js
import mime from "mime/lite.js"
import { scripts, styles } from "./extra.js"
import { routes, get } from "./routes.js"
import { toHeaders } from "./utils.js"

/**
 * generate html
 * @param {import("./private.js").HtmlOptions} options options 
 * @returns {Promise<import("./private.js").HtmlResponse>}
 */
export default async function html ({ dir, path, wsPort, statik, listDir }) {
	// hacky workaround, opened an issue on cheap-watch
	if (!statik && (!routes.has(path) || path == "/")) {
		const is = await get(dir, path.slice(1))
		if (!is) return { code: 404 }
	} else if (statik && !routes.has(path)) {
		return { code: 404 }
	}

	const route = /** @type {import("./private.js").Route} */ (routes.get(path))
	if (route.directory) {
		const { children } = route

		if (!listDir) {
			const index = children.filter(({ directory }) => !directory).find(({ path }) => path == "index.html" || path == "index.htm")
			if (index) return html({ dir, path: `${path}/${index.path}`.replace(/\/{2,}/g, "/"), wsPort, statik })
		}

		const content = doDirectory({ children, path, wsPort })
		const headers = toHeaders({ content, type: "text/html" })

		return {
			code: 200,
			content,
			headers
		}
	} else {
		let { content } = route
		const type = mime.getType(path) || "text/plain"

		switch (type) {
			case "text/html": {
				if (!statik) content = livereload({ content, wsPort })
				const headers = toHeaders({ content, type })

				return {
					code: 200,
					content,
					headers
				}
			}
			default: {
				const headers = toHeaders({ content, type })

				return {
					code: 200,
					content,
					headers
				}
			}
		}
	}
}

/**
 * inject live-reload script
 * @param {{ content: Buffer, wsPort: number }} options options
 * @returns {Buffer}
 */
function livereload ({ content, wsPort }) {
	const script = Buffer.from(scripts.livereload(wsPort))
	const contentBuffer = Buffer.from(content)
	return Buffer.concat([ contentBuffer, script ])
}

/**
 * generate directory html
 * @param {{ children: { path: string, directory: boolean }[], path: string, wsPort: number }} options options 
 * @returns {Buffer}
 */
function doDirectory ({ children, path, wsPort }) {
	const dir = path.slice(1).split(/\/+/g).filter(( dir ) => !!dir.length).map(( dir, i, pre ) => ({ name: dir, link: `/${pre.slice(0, i + 1).join("/")}` }))
	const links = children.map(({ path: p, directory }) => `<a href="${encodeURI(`/${path}/${p}`.replace(/\/+/g, "/"))}">${p}${directory ? "/" : ""}</a>`)

	return Buffer.from([
		"<!DOCTYPE html>",
		"<html>",
		"<head>",
			`<title>${dir[dir.length - 1]?.name || "/"}</title>`,
			scripts.livereload(wsPort),
			styles.directory,
		"</head>",
		"<body>",
		"<div class=\"main\">",
			"<div class=\"name\">",
				"<a href=\"/\">./</a>",
				dir.map(({ name, link }) => `<a href="${link}">${name}/</a>`).join(""),
			"</div>",
			"<div class=\"entries\">",
				links.join("\n"),
			"</div>",
		"</div>",
		"</body>",
		"</html>"
	].join(""))
}
